### ISSUE TITLE
BUG: backwards compatibility break in master with distribution broadcasting

### ISSUE THREAD

**Issue by josef-pkt** [scipy/scipy#13675]
see #13589

AFAICS, #13611 broke statsmodels code, change is not backwards compatible

https://github.com/statsmodels/statsmodels/issues/7371

statsmodels uses a subclass of rv_discrete

we have our own copy of scipy's lazywhere
AFAICS it breaks with scipy master because elements don't match in shape anymore

```
    def _logpmf(self, x, mu, w):
        return _lazywhere(x != 0, (x, mu, w),
                          (lambda x, mu, w: np.log(1. - w) + x * np.log(mu) -
                          gammaln(x + 1.) - mu),
                          np.log(w + (1. - w) * np.exp(-mu)))
```

the final part of the traceback in azur testing is

```
>       return _insert(arr, mask, vals)
E       ValueError: place: mask and data must be the same size

/opt/hostedtoolcache/Python/3.9.2/x64/lib/python3.9/site-packages/numpy/lib/function_base.py:1748: ValueError
```

**Comment by iandall** [comment]
This specific issue seems to be because statsmodels is using the current scipy but an old copy of `_lazywhere`.

It is possible that other user's sub-classes could be broken. I have a branch where I was working on an opt in model (preserving the old argsreduce by default) which I could clean up and submit if there is interest. It does mean a significant amount of cruft in the code which might not be considered worth it.

**Comment by rgommers** [comment]
Thanks for the report @josef-pkt, and for looking into it @iandall. 

It seems kind of painful to break the latest released statsmodels when we release SciPy 1.7.0. There's also a second regression (already fixed) reported on the PR: https://github.com/scipy/scipy/pull/13611#issuecomment-812900208. So if we know that there are likely to be more cases of broken subclasses, we may have to consider the opt-in model (which admittedly isn't great). I'm not quite sure, so labelling with `1.7.0` so we won't forget.

Perhaps we should see if there's a workaround for statsmodels, and leave it as is in `1.7.0rc1` and see if anyone else notices?

**Comment by josef-pkt** [comment]
It's fixed in statsmodels master using scipy's `_lazywhere` instead of our old copy  https://github.com/statsmodels/statsmodels/pull/7385

I don't think there is a workaround for released statsmodels except monkey patching or replacing statsmodels' lazywhere

**Comment by rgommers** [comment]
Thanks @josef-pkt 

> It is possible that other user's sub-classes could be broken. I have a branch where I was working on an opt in model (preserving the old argsreduce by default) which I could clean up and submit if there is interest. It does mean a significant amount of cruft in the code which might not be considered worth it.

We may need to do this then. Cc @WarrenWeckesser who found the other regression, and @ev-br. Opinions?

**Comment by ev-br** [comment]
Is there a way to retain improvements in a back compatible way?

**Comment by ev-br** [comment]
A standalone test MWE would help, too

**Comment by josef-pkt** [comment]
The test case in statsmodels requires the old scipy version of lazywhere.
For inplace modification of arrays, AFAICS, there are many cases in the original scipy code https://github.com/scipy/scipy/pull/13611/files

If it's opt-in, then all scipy distributions can opt-in and get the speed advantage.
This would still break code for subclasses of the specific distributions, but my guess is that there are few user cases like that.

**Comment by ev-br** [comment]
So your suggestion is then to document the breaking change in the release notes and see how bad the breakage is, Josef?

**Comment by tylerjereddy** [comment]
@mdhaber I'm not likely going to add a release note entry for this this evening, but it may be prudent to do so for `stats`  backwards incompatible changes during the release candidate phase perhaps..

**Comment by mdhaber** [comment]
@tylerjereddy OK. I don't currently know about this issue and I can't think of any other backwards-incompatible changes.
@rgommers can you add this, since it looks like you worked on gh-13611?

**Comment by rgommers** [comment]
Guess we didn't get around to this - let's see if anyone complains......

**Comment by rgommers** [comment]
No one else complained, and it was fixed in `statsmodels` with its Oct'21 `0.13.0` release. It's not likely there are other (high-profile) users, because this relied on private API (`lazywhere` from `scipy._lib`). So I propose we close this issue.
